corehq/apps/app_manager/views/utils.py

```diff
@@ -352,7 +352,10 @@ def update_linked_app(app, master_app_id_or_build, user_id):
         or master_build.version > previous.upstream_version
         or toggles.MULTI_MASTER_LINKED_DOMAINS.enabled(app.domain)
     ):
-        old_multimedia_ids = set([media_info.multimedia_id for path, media_info in app.multimedia_map.items()])
+        old_multimedia_ids = set(
+            [getattr(media_info, 'upstream_media_id', None) or media_info.multimedia_id
+             for path, media_info in app.multimedia_map.items()]
+        )
```

Comment:
Why is this change necessary? I think `old_multimedia_ids` is only used to determine which media is missing from the local server in which case it should only need the local ID and not the upstream ID. I think using the upstream ID will create false positives for missing media.

Reply:
I found that "old" here means the collection of multimedia already present in the older version of the downstream app, before the newer app version is pulled. [_get_missing_multimedia](https://github.com/dimagi/commcare-hq/pull/32797/files#diff-b565707cd8670d0f76525035ea595b0d2617a8d87e5575e5273d945f33e6468eR123-R135) ultimately checks the ids from the newly pulled multimedia map against `old_multimedia_ids`, and if anything is missing from `old_multimedia_ids`, adds it to a list to `fetch_remote_media`. 

Essentially I made this change _because_ I was getting false positives when looking for updated multimedia ids in a list containing only the local ids, resulting in a new `AuxMedia` entry on every app pull.

Reply:
Well, you were correct that this did have an unintended consequence. On repeated app pulls (particularly with the `MULTI_MASTER` feature flag enabled) the multimedia map is not being rewritten from its upstream version, because that happens in `fetch_remote_media` which is never called if the multimedia does not appear to be "missing" in the first place.

Reverting this line is a clear way to fix the issue, but then a new `AuxMedia` entry gets created on the `CommCareMultimedia` object on every app pull (where there is media that conflicts between upstream & downstream). I assume this isn't ideal, but how big a concern would it be, really?

Reply:
I looked at this again and think your original change was correct in terms of pulling missing MM. The issue you describe with not updating the multimedia map seems like a separate issue.

I think we should look at updating the `corehq.apps.app_manager.views.utils.overwrite_app` method to do a smarter conversion of the MM map rather than just overwriting the local version with the remote version. Bascially we want to preserve the 'upstream_id' in the media items. I think then we'd be able to correctly determine what's actually missing without having to re-fetch anything.

Reply:
@snopoke I took a stab at this today, let me know if it looks right. Now, it's still `fetch_remote_media` that's responsible to set the correct ids for the local and the upstream reference, because it uses the actual file data/hash to determine the media already exists locally. `overwrite_app` is now also comparing the existing and incoming multimedia maps, and if an item in the existing map has the same file path plus an `upstream_media_id` that matches the incoming `multimedia_id`, it keeps the existing multimedia map item instead. 

There's one minor refactor included as well, setting the incoming multimedia map's versions to `None` during comparison rather than in a separate loop.